source/base/data_out_base.cc

```diff
@@ -5597,43 +5622,80 @@ namespace DataOutBase
-                  data.push_back(0);
-                  data.push_back(0);
-                  break;
+                switch (last_component - first_component)
+                  {
+                    case 0:
+                      data.push_back(data_vectors(first_component, n));
+                      data.push_back(0);
+                      data.push_back(0);
+                      break;
+
+                    case 1:
+                      data.push_back(data_vectors(first_component, n));
+                      data.push_back(data_vectors(first_component + 1, n));
+                      data.push_back(0);
+                      break;
+
+                    case 2:
+                      data.push_back(data_vectors(first_component, n));
+                      data.push_back(data_vectors(first_component + 1, n));
+                      data.push_back(data_vectors(first_component + 2, n));
+                      break;
+
+                    default:
+                      // Anything else is not yet implemented
+                      Assert(false, ExcInternalError());
+                  }
+              }
+            else
+              {
+                Tensor<2, 3> vtk_data;
+                vtk_data = 0.;
 
-                case 1:
-                  data.push_back(data_vectors(first_component, n));
-                  data.push_back(data_vectors(first_component + 1, n));
-                  data.push_back(0);
-                  break;
-                case 2:
-                  data.push_back(data_vectors(first_component, n));
-                  data.push_back(data_vectors(first_component + 1, n));
-                  data.push_back(data_vectors(first_component + 2, n));
-                  break;
+                const unsigned int size = last_component - first_component + 1;
+                if (size == 1)
+                  // 1D, 1 element
+                  {
+                    vtk_data[0][0] = data_vectors(first_component, n);
+                  }
+                else if ((size == 4) || (size == 9))
+                  // 2D, 4 elements or 3D 9 elements
+                  {
+                    for (unsigned int c = 0; c < size; ++c)
+                      {
+                        const auto ind =
+                          internal::unrolled_to_component_indices<dim>(c);
+                        vtk_data[ind[0]][ind[1]] =
+                          data_vectors(first_component + c, n);
+                      }
+                  }
+                else
+                  {
+                    Assert(false, ExcInternalError());
+                  }
 
-                default:
-                  // VTK doesn't support anything else than vectors with 1, 2,
-                  // or 3 components
-                  Assert(false, ExcInternalError());
+                // now put the tensor into data
+                for (unsigned int i = 0; i < 3; ++i)
+                  for (unsigned int j = 0; j < 3; ++j)
```

Comment:
Can you add a comment here saying that you are padding with zeros because VTK format *always* wants to see a 3x3 tensor, regardless of dimension?